### ISSUE TITLE
Activity categories MiBand2/BIP

### ISSUE THREAD

**Issue by weliem** [Freeyourgadget/Gadgetbridge#1030]
Hi folks,

Great work here! I thought I contribute something. I have been studying the category codes that come from the MiBand2 and BIP. What GadgetBridge currently does is not correct since only the lower 4 bits are masked which shouldn't be done. 

Here is what I have found so far:

Walking = 1,
Running = 98,
Charging = 6,
SlowWalking = 16,
WakeUp = 28,
StairClimbing = 33,
Sitting = 80,
NotWornFaceUp = 83,
SedentaryInLast5Minutes = 90
Standing = 96,
SleepInterruption = 105,
Sleeping = 112,
NotWornFaceDown = 115,
SleepUnknown1 = 121,
SleepUnknown2 = 122,
SleepStart = 123,
SleepUnknown3 = 124,

Some notes on sleep:
- The device only has the category 'sleeping', there is no light/deep sleep
- Deep sleep is characterized by prolonged periods of 0 intensity. It seems that after +/- 10 minutes of 0 intensity deep sleep kicks in and you go back to light sleep if intensity goes up again to > 15.
- After some sleep (112) states a SleepStart (123) is usually present but not always. 
- Sleep is ended by WakeUp (28) but not always....

But I think it is really interesting to see it actually does sitting, standing, stairclimbing etc! These values are not used in the Mi Fit app but I am 100% sure about them!

**Comment by walkjivefly** [comment]
Which firmware and hardware versions?

**Comment by weliem** [comment]
MiBand2 : Firmware 1.0.1.69, Hardware 0.1.3.2
BIP : Firmware 0.1.0.86, Hardware 0.11.2.4

**Comment by ashimokawa** [comment]
@weliem 
Thank you for the information.

Did you read #686 ?
I was assuming the upper bits are newly introduced flags that help postprocessing and sanitizing the data from the device.
We had a lot of assumption what these "flags" could mean and were discussing it while looking at the great table @Roxxor91 provided. But for every assumption there was some example that did not fit. You could also download it at see if your assumption regarding deep sleep fits.

Anyway the details you mentioned here are enough to justify a re-evaluation of everything ;) Thanks!

**Comment by weliem** [comment]
Yes, I saw issue #686 . @Roxxor91 had some of them right already so I had another look and expanded the list a bit...

**Comment by LuccoJ** [comment]
Will these be meaningful, though, unless you have "trained" those activities first? That's possible to do on the Mi Fit app... I've never tried it because it seems *extremely* intrusive, as I know Huami have some API for third-parties to, pretty much, know when I'm brushing my teeth... and I am the one teaching them how to recognize that!

But if Gadgetbridge could be made to both read the flags *and* initiate training to let the watch learn to recognize these activities, it would be a great addition. I didn't really think this was done within the watch, I had assumed it was post-processing that Mi Fit did.

**Comment by weliem** [comment]
@LuccoJ not sure what you mean with 'training'. Certainly never did anything like that....these values come 'out of the box'....

But of course feel free to verify my findings!

**Comment by LuccoJ** [comment]
@weliem I am referring to [this functionality in Mi Fit](https://www.flickr.com/photos/ljlbox/40097159785/in/dateposted-public/), where those things like stair climbing (but also additional ones like brushing teeth) are termed "Behavior tagging" which you train with "data collection" that gets "uploaded".

Maybe it's an entirely different thing, but the overlap confuses me.

**Comment by weliem** [comment]
@LuccoJ , I see it now, thanks. However, I never used that since I didn't even know it existed....

**Comment by pm-cz** [comment]
@LuccoJ I don't think the activity learning is related to the band/watch itself. It is used for activity tagging, probably with raw accelerometer access endpoint, which is then sent to Huami to improve the future versions of firmware detection (this is the upload, I think, as it requires data connection).
